libsoftwarecontainer/src/softwarecontainer.cpp

```diff
@@ -165,30 +165,92 @@ void SoftwareContainer::addGateway(Gateway *gateway)
     m_gateways.push_back(std::move(std::unique_ptr<Gateway>(gateway)));
 }
 
-void SoftwareContainer::setGatewayConfigs(const GatewayConfiguration &gwConfig)
+ReturnCode SoftwareContainer::setGatewayConfigs(const GatewayConfiguration &gwConfig)
 {
-    // Go through the active gateways and check if there is a configuration for it
-    // If there is, apply it.
+    ReturnCode result = ReturnCode::SUCCESS;
 
+    result = configureGateways(gwConfig);
+    if (isError(result)) {
+        return result;
+    }
+
+    result = activateGateways();
+    if (isError(result)) {
+        return result;
+    }
+
+    m_containerState.setValueNotify(ContainerState::READY);
+
+    return result;
+}
+
+ReturnCode SoftwareContainer::configureGateways(const GatewayConfiguration &gwConfig)
+{
     for (auto &gateway : m_gateways) {
         std::string gatewayId = gateway->id();
 
         json_t *config = gwConfig.config(gatewayId);
         if (config != nullptr) {
             char *configStr = json_dumps(config,0);
```

Comment:
I'm guessing we don't ever want to use this as a `char *`, so you could implicitly convert it to a string by just using std::string as the type. It will be freed when the string goes out of scope anyway.

Then we'd get rid of the conversion below.